ffi/src/lib.rs

```diff
@@ -1245,9 +1246,55 @@ pub unsafe extern "C" fn tx_report_entity_for_temp_id(tx_report: *mut TxReport,
     }
 }
 
-// TODO: cache
+/// Adds an attribute to the cache.
+/// `store_cache_attribute_forward` caches values for an attribute keyed by entity
+/// (i.e. find values and entities that have this attribute, or find values of attribute for an entity)
+///
+/// # Safety
+///
+/// Callers must ensure that the pointer to the `Store` is not dangling and that
+/// the C string provided to `attribute` is valid.
+#[no_mangle]
+pub extern "C" fn store_cache_attribute_forward(store: *mut Store, attribute: *const c_char) -> *mut ExternResult {
+    let store = unsafe { &mut *store };
+    let kw = kw_from_string(c_char_to_string(attribute));
+    Box::into_raw(Box::new(store.cache(&kw, CacheDirection::Forward).into()))
+}
+
+/// Adds an attribute to the cache.
+/// `store_cache_attribute_reverse` caches entities for an attribute keyed by value.
+/// (i.e. find entities that have a particular value for an attribute).
+///
+/// # Safety
+///
+/// Callers must ensure that the pointer to the `Store` is not dangling and that
+/// the C string provided to `attribute` is valid.
+#[no_mangle]
+pub extern "C" fn store_cache_attribute_reverse(store: *mut Store, attribute: *const c_char) -> *mut ExternResult {
+    let store = unsafe { &mut *store };
+    let kw = kw_from_string(c_char_to_string(attribute));
```

Comment:
Again, similar comments to the earlier (documentation) PR -- can we avoid this allocation?  Maybe we need to figure out how to find internal string references so that we can look up `NamespacedKeyword` by `(&str, &str)` (and therefore, hopefully, by `(CStr, CStr)`).  That is, can we make `(&str, &str) impl Borrow<&NamespacedKeyword>`?  I don't know if that's possible without changing the representation of `NamespacedKeyword`.  But it's going to be very common, so we might want to do this.